libraries/AP_Relay/AP_Relay.cpp

```diff
@@ -128,61 +111,167 @@ AP_Relay::AP_Relay(void)
     singleton = this;
 }
 
-
-void AP_Relay::init()
+void AP_Relay::convert_params()
 {
-    if (_default != 0 && _default != 1) {
+    // bail immediately if we have already done this
+    if (_params[0].pin.configured_in_storage ()) {
         return;
     }
-    for (uint8_t i=0; i<AP_RELAY_NUM_RELAYS; i++) {
-        set(i, _default);
+
+    int8_t default_state = 0; // off was the old behaviour
+
+    // grab the old values if they were set
+    for (uint8_t i = 0; i < 4; i++) {
+        int8_t pin = 0;
+        if (AP_Param::get_param_by_index(this, i, AP_PARAM_INT8, &pin)) {
+            // if the pin isn't negative we can assume the user might have been using it, map it to the old school relay interface
+            _params[i].pin.set_and_save(pin);
+            _params[i].function.set_and_save(1);
+        }
+    }
+    for (uint8_t i = 4; i < 6; i++) { // continuation of the above loop, we just have to step over the old default
+        int8_t pin = 0;
+        if (AP_Param::get_param_by_index(this, i + 1, AP_PARAM_INT8, &pin)) {
+            // if the pin isn't negative we can assume the user might have been using it, map it to the old school relay interface
+            _params[i].pin.set_and_save(pin);
+            _params[i].function.set_and_save(1);
+        }
     }
+
+    const bool asserted_default = AP_Param::get_param_by_index(this, 4, AP_PARAM_INT8, &default_state);
+    for (uint8_t i = 0; i < MIN(AP_RELAY_NUM_RELAYS, 6); i++) {
+        if (asserted_default) {
+            _params[i].default_state.set_and_save(default_state);
+        }
+    }
+
+    _params[0].pin.save (true);
 }
 
-void AP_Relay::set(const uint8_t instance, const bool value)
+void AP_Relay::set_defaults () {
+    const int8_t pins[6] = { RELAY1_PIN_DEFAULT,
+                             RELAY2_PIN_DEFAULT,
+                             RELAY3_PIN_DEFAULT,
+                             RELAY4_PIN_DEFAULT,
+                             RELAY5_PIN_DEFAULT,
+                             RELAY6_PIN_DEFAULT };
+
+    for (uint8_t i = 0; i < MIN(AP_RELAY_NUM_RELAYS, 6); i++) {
+        // set the default
+        if (pins[i] != -1) {
+            _params[i].pin.set_and_default(pins[i]);
+        }
+    }
+}
+
+void AP_Relay::init()
 {
-    if (instance >= AP_RELAY_NUM_RELAYS) {
-        return;
+    set_defaults ();
```

Comment:
nitpick: space between function name and brackets.

Reply:
fixed